File: Algorithms/C++/00724-Find Pivot Index.cpp
class Solution {
public:
    int pivotIndex(vector<int>& nums) {
        int right{0};
        for(int i{0}; i < nums.size(); i++){
            right += nums[i];
        }

        int left{0};
        int pivot{0};
        while(pivot < nums.size()){
            right -= nums[pivot];
            if(left == right) return pivot;
            left += nums[pivot];
            pivot++;
        }

        return -1;
    }
};

/*
    tc: O(n)
    sc: O(1)

    [-1,-1,-1,-1,-1,0]
    left = -1
    right = -3
    pivot = 1

    while( left < right)
        if(left == right) return index

    return -1

    pass 1: sum entire array
    pass 2: iterate pivot from left to right to find index

*/

File: Algorithms/Python3/00724-Find Pivot Index.py
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        total = sum(nums)
        leftSum = 0
        for i, pivot in enumerate(nums):
            rightSum = total - pivot - leftSum
            if leftSum == rightSum:
                return i
            leftSum += pivot
        
        return -1

'''
[1,7,3,6,5,6]
leftSum = 1
rightSum = total - pivot - leftSum = 27
28 - 7 - 1 = 20
sum = 28

tc: O(n)
sc: O(1)

'''